File: C/AL-assignment3.c
// #include <iostream.h>
#include <limits.h>
#include <math.h>
#include <stdbool.h>
#include <stdio.h>

#define MAX 1000

// get distance function
float distance(float a[2], float b[2]) { return sqrt(pow(a[0] - b[0], 2) + pow(a[1] - b[1], 2)); }

float min_distance(float points[MAX][2], int n) {
    float res = 0.0;
    int used = 0;

    bool check[n];
    float minDist[n];

    for (int i = 0; i < n; i++) {
        minDist[i] = INT_MAX;
        check[i] = false;
    }
    minDist[0] = 0;

    while (used < n) {
        float min_dst = INT_MAX;
        int cur = -1;
        for (int i = 0; i < n; ++i) {
            if (!check[i] && min_dst > minDist[i]) {
                min_dst = minDist[i];
                cur = i;
            }
        }
        res += min_dst;
        used++;
        check[cur] = true;
        for (int i = 0; i < n; i++) {
            float dst = distance(points[cur], points[i]);
            if (!check[i] && minDist[i] > dst) {
                minDist[i] = dst;
            }
        }
    }
    return res;
}

int main() {
    float array[3][2] = {
        {1, 2},   // row 0 = 1, 2, 0, 0, 0
        {6, 7},   // row 1 = 6, 7, 8, 0, 0
        {11, 12}  // row 2 = 11, 12, 13, 14, 0
    };
    float res = min_distance(array, 3);
    printf("res : %f\n", res);
    return 0;
}

File: python/AL-assignment3.py
def distance(a, b):
    return ((a[0] - b[0]) ** 2 + (a[1] - b[1]) ** 2) ** (0.5)

def min_points(mst, points, used):
    dis = -1
    ind = -1
    for p in mst:
        for i in range(len(points)):
            if used[i] == 0:
                q = points[i]
                temp = distance(p, q)
                if dis == -1 or dis >= temp:
                    dis = temp
                    ind = i
                    # print("temp : ", temp, " i  : ", i)
    # print("dis : ", dis)
    # print("ind : ", ind)
    mst.append(points[ind])
    used[ind] = 1
    return dis

def min_distance(points, n):
    d = 0.0
    arr = [points[0]]
    used = [0] * n
    used[0] = 1
    for i in range(n - 1):
        dis = min_points(arr, points, used)
        print(used)
        d += dis
    return d

test = [
    [1, 2],
    [3, 4],
    [5, 6],
    [7, 8],
]

test2 = [
    [3, 12],
    [-2, 5],
    [-4, 1],
    [7, 9],
]

res = min_distance(test2, len(test2))
print(res)
# print(distance(test[0], test[1]))

